File: arrays-strings/05-longest-common-prefix.py
"""
Problem 05: Longest Common Prefix (LeetCode #14)
Topic: Arrays & Strings
Difficulty: Easy-Medium

Description:
Write a function to find the longest common prefix string amongst an array of strings.
If there is no common prefix, return an empty string "".

Time Complexity: O(S) where S is the total number of characters in all strings
Space Complexity: O(1)
"""

from typing import List

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if not strs:
            return ""
        
        prefix = strs[0]
        for i in range(1, len(strs)):
            while not strs[i].startswith(prefix):
                prefix = prefix[:-1]
                if not prefix:
                    return ""
        return prefix

if __name__ == "__main__":
    sol = Solution()
    print("Test 1:", sol.longestCommonPrefix(["flower", "flow", "flight"]))  # Expected: "fl"
    print("Test 2:", sol.longestCommonPrefix(["dog", "racecar", "car"]))     # Expected: ""

File: arrays-strings/05-longest-common-prefix.c
/*
 * Problem 05: Longest Common Prefix (LeetCode #14)
 * Topic: Arrays & Strings
 * Difficulty: Easy-Medium
 * Language: C
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char* longestCommonPrefix(char** strs, int strsSize) {
    if (strsSize == 0) return "";
    
    char* prefix = (char*)malloc(strlen(strs[0]) + 1);
    strcpy(prefix, strs[0]);
    
    for (int i = 1; i < strsSize; i++) {
        int j = 0;
        while (prefix[j] && strs[i][j] && prefix[j] == strs[i][j]) {
            j++;
        }
        prefix[j] = '\0';
        if (prefix[0] == '\0') break;
    }
    return prefix;
}

int main() {
    char* strs1[] = {"flower", "flow", "flight"};
    char* res1 = longestCommonPrefix(strs1, 3);
    printf("Test 1: \"%s\"\n", res1); // Expected: "fl"
    free(res1);

    char* strs2[] = {"dog", "racecar", "car"};
    char* res2 = longestCommonPrefix(strs2, 3);
    printf("Test 2: \"%s\"\n", res2); // Expected: ""
    free(res2);
    return 0;
}
